backends/ref/ceed-ref-restriction.c

```diff
@@ -70,67 +68,43 @@ static int CeedElemRestrictionApply_Ref(CeedElemRestriction r,
     // Note: in transpose mode, we perform: v += r^t * u
     esize = (nblk - 1)*blksize*elemsize;
     if (!impl->indices) {
-      for (CeedInt shift=0; shift<nblk*blksize*ncomp*elemsize;
-           shift+=blksize*ncomp*elemsize) {
-        CeedInt maxj = (shift<(nblk-1)*blksize*ncomp*elemsize)?blksize:r->nelem%nblk;
-        if (maxj == 0) maxj = blksize;
-        for (CeedInt j = 0; j<maxj; j++) {
-          for (CeedInt k = 0; k<ncomp*elemsize; k++) {
-            vv[shift + j*ncomp*elemsize + k] = uu[shift + k*blksize + j];
+      for (CeedInt e = 0; e < nblk*blksize; e+=blksize) {
+        CeedInt maxj = ((e<(nblk-1)*blksize)||!(r->nelem%blksize))?blksize:r->nelem%blksize;
+        for (CeedInt j = 0; j < maxj; j++) {
+          for (CeedInt k = 0; k < ncomp*elemsize; k++) {
+            vv[(e + j)*ncomp*elemsize + k] += uu[e*ncomp*elemsize + k*blksize + j];
           }
         }
       }
     } else if (ncomp == 1) {
-      for (CeedInt i = 0; i<esize; i++) vv[impl->indices[i]] += uu[i];
-      CeedInt nlastelems = r->nelem % blksize;
+      for (CeedInt i = 0; i < esize; i++) vv[impl->indices[i]] += uu[i];
+      CeedInt nlastelems = (r->nelem%blksize)?r->nelem%blksize:blksize;
       CeedInt shift = (nblk - 1)*blksize*elemsize;
-      if (nlastelems == 0) nlastelems = blksize;
-      for (CeedInt i = 0; i<blksize*elemsize; i++) {
+      for (CeedInt i = 0; i < blksize*elemsize; i++) {
         if ((i % blksize) < nlastelems) {
           vv[impl->indices[shift + i]] += uu[shift + i];
         }
       }
     } else {
       // u is (elemsize x ncomp x nelem)
       if (lmode == CEED_NOTRANSPOSE) { // vv is (ndof x ncomp), column-major
```

Comment:
I am suspicious of my logic here after the loop reworking and want to look at again in the morning.